udemy_exercises/cell_phone/CellPhone.java

```diff
@@ -1,9 +1,211 @@
+        }
+    }
+
+    public static void main(String[] args) {
+        boolean quit = false;
+        startPhone();
+        printActions();
+        while (!quit) {
+            System.out.println("\nEnter action: (6 to show available actions)");
+            int action = scanner.nextInt();
+            scanner.nextLine();
+            switch (action) {
+                case 0:
+                    cellPhone.quit();
+                    quit = true;
+                    break;
+                case 1:
+                    cellPhone.showContacts();
+                    break;
+                case 2:
+                    addNewContact();
+                    break;
+                case 3:
+                    updateContact();
+                    break;
+                case 4:
+                    removeContact();
+                    break;
+                case 5:
+                    searchContact();
+                    break;
+                case 6:
+                    printActions();
+                    break;
+            }
+        }
+    }
+
+    private static void addNewContact() {
+        System.out.println("Enter new contact name: ");
+        String name = scanner.nextLine();
+        System.out.println("Enter new phone number: ");
+        String phone = scanner.nextLine();
+        cell_phone.Contact newContact = cell_phone.Contact.saveContact(name, phone);
+        if (cellPhone.addContact(newContact)) {
+            System.out.println("New contact added: " + name + " -> " + phone);
+        } else {
+            System.out.println("Cannot add, " + name + " is already on file.");
+        }
+    }
+
+    private static void updateContact() {
+        System.out.println("Enter existing contact name: ");
+        String name = scanner.nextLine();
+        cell_phone.Contact existingContactRecord = cellPhone.searchContact(name);
+        if (existingContactRecord == null) {
+            System.out.println("Contact not found.");
+            return;
+        }
+        System.out.println("Enter new contact name: ");
+        String newName = scanner.nextLine();
+        System.out.println("Enter new contact phone number: (leave blank if you do not wish to change the number");
+        String newNum = scanner.nextLine();
+        //users can change only the name by leaving the phone number blank
+        cell_phone.Contact newContact;
+        if (newNum.equals("")) {
+            newContact = cell_phone.Contact.saveContact(newName, existingContactRecord.getPhoneNumber());
+        } else {
+            newContact = cell_phone.Contact.saveContact(newName, newNum);
+        }
+        if (cellPhone.updateContact(existingContactRecord, newContact)) {
+            System.out.println("Successfully updated contact.");
+        } else {
+            System.out.println("Error updating contact");
+        }
```

Comment:
Good encapsulation of these details to keep them out of the 'menu' code, but I think they are a bit more complex than is necessary.  They serve the purpose, but the complexity introduces risk.